### ISSUE TITLE
Extracted fern mesh with appearance

### ISSUE THREAD

**Issue by RuiYangQuan** [qway/nerfmeshes#18]
hi,thank you very much for your work, but I have encountered a problem here, when I generate the mesh of fern, I get this kind of graph, what is this situation, I would appreciate your advic!!!
![3](https://user-images.githubusercontent.com/90050796/182093836-2ce55a39-fc99-425d-9181-1f995ad2490f.png)
e

**Comment by DomainFlag** [comment]
Hi @RuiYangQuan, I've checked the model on my regular laptop and it seems there is a problem with the parameters (most likely). Will check throughly next week when getting back to my workstation.

**Comment by RuiYangQuan** [comment]
Okay, thank you for your reply. If you find anything new, please let me know immediately. @DomainFlag

**Comment by DomainFlag** [comment]
Hi @RuiYangQuan, I did some checking, and it seems that the uploaded model has some issues (either was not properly uploaded or alike). I don't have past models locally, thus, I will try re-do the training and slightly make the project up-to-date. Will try complete it by the end of this week, as I am running other experiments for other projects as well. Thank you a lot for submitting the issue.

**Comment by RuiYangQuan** [comment]
> @DomainFlag 

Hi,i wonder like to know if this problem has been solved.Could you please provide the latest pre-training model.

**Comment by DomainFlag** [comment]
Hi @RuiYangQuan, I've actually did some testing, and there wasn't an issue with the model.

 The actual problem is the data which is used for reconstructing the scene, or I would better say the limitations of the model itself when trained on that: 

If you take LLFF as an example, the poses from which the photos were taken are not spherically uniformly distributed as in the Synthetic dataset. That results in unresolved scene mass which is not any useful for scene reconstruction if applying Marching Cubes algorithm. For that case we have  `mesh_surface_ray.py` which does not solely rely on the scene density inferred from the trained parametric function, but uses ray-casting and triangulation on the fly. But, we didn't have enough time to fully port it for production use-cases.

Also note: the BuFF model is suitable for sparse scenes and closed ones, because the volumetric encoded scene will be sparse as well, thus that's where the gain comes from.

I did some changes, and trained a few more models on (Fern, Room) and did some changes to `mesh_surface_ray.py` but will need more time. I will push this weekend whatever is workable but don't expect a full solution as I am jointly working on another paper that has a deadline soon.

**Comment by wjlszhang** [comment]
Hi, @DomainFlag 
Have you solved the issue? I'm not sure it is cause by the same reason.
When I use mesh_nerf.py to generate mesh.obj, it usually takes long time and the size of mesh.obj is purly large, i.e. around 20G

Notes: 
Firstly I use colmap to generate my own dataset as suggested, and it is dental related data, one of the pictures is as bellow  
![image](https://user-images.githubusercontent.com/101317599/199644814-5db83283-17db-4c28-bcd0-36ec48b0015d.png)
And then use rain_nerf.py to train the model, the result looks good 
![1485c55db9fcb17563b5c632e312a11](https://user-images.githubusercontent.com/101317599/199644880-8a7e46f8-91ed-486a-8c11-bc24735822aa.jpg)
Last, use mesh_nerf.py to generate mesh obj, as blow
python src/mesh_nerf.py --log-checkpoint logs/nerf-colmap-fern/default/version_18 --checkpoint model_last.ckpt --save-dir ./meshdata --limit 1.2 --res 480 --iso-level 32 --view-disparity-max-bound 1e0

Thanks much

**Comment by DomainFlag** [comment]
@RuiYangQuan, @wjlszhang

I will try to clear things up with the following comment, as it seems there is some misunderstanding although the comment will require some additional re-editing. Feel free to jump to the solution part if needed, however, it might not be the one based on your requirements.

## Problem

NeRF and all its equivalents have no prior knowledge to work with modeling objects in the scene (not to say dealing with scene understanding or other variants alike). It's like asking someone who hasn't seen a chair in his whole life to model it by only having the front view of it with the other 2 legs occluded. Of course, a model like NeRF won't be able to model it properly and you'll end up with a chair with missing legs at best. That's the reason you can't extract a mesh from the examples above, it's just there is no prior knowledge to work with, although it can be done something (some great progress is done recently for generative methods in 2D). However, that won't be called NeRF anymore (most likely).

It's important to set the expectations right about the mesh to be extracted - "what do we expect based on what is given". As an example, I've added one such example (**Room** from **LLFF** dataset), just to clear up the confusion. If having insufficient imagery, you won't even get the sort-of billboards (e.g., back of the seats) that you can find in 3D game development as compensation but some sort of interpolated mess instead. Additionally, the model is overly reliant on view dependence (no 360deg imagery available, all shots are done from the front of the room) from the train set as it fails to recover the back of the scene. No such information is available in the training data thus you end up with poor **local** generalization. Even so, a slight perturbation applied to the GT ray origins can drastically affect the results. Note, that may not happen if considering all the mentioned points at the bottom's solution. The model isn't able to handle such constraints, unfortunately, thus you end up with ill-defined parametrization of the space (density) which is not any good for something like Marching Cubes.

A single reference image - **Room** from **LLFF** Dataset ([Link](https://drive.google.com/drive/folders/128yBriW1IG_3NJ5Rp7APSTZsJqdJdfc1))

Pre-trained model PSNR (34.65) on test set:

## A few renders:
| Example 1 | Example 2 | Example 3 |
| -------------- | -------------- |  -------------- | 
| ![0000](https://user-images.githubusercontent.com/23214006/200189855-8778814c-8962-47f2-8130-4cc1466704b6.png) | ![0001](https://user-images.githubusercontent.com/23214006/200189856-ec0e9e07-9bf8-4869-aa91-2d22324a6ca6.png) | ![0002](https://user-images.githubusercontent.com/23214006/200189858-6cccc47a-d7ff-4eb1-bbcc-87f8fb680693.png) |
| ![0000](https://user-images.githubusercontent.com/23214006/200189866-c527b7b6-0a3e-45b1-9e86-f6623d9f368b.png) | ![0001](https://user-images.githubusercontent.com/23214006/200189867-f3bacc88-ab0d-4392-9631-8d80bef38902.png) | ![0002](https://user-images.githubusercontent.com/23214006/200189868-b87ecf11-3779-43ac-8964-0d19baed4f51.png) |

Ray-casting based on poses uniformly sampled across a sphere (360deg):

| Example 1 | Example 2 |
| -------------- | -------------- |
| ![Screenshot from 2022-11-06 15-22-27](https://user-images.githubusercontent.com/23214006/200180073-0d5ee3a9-7fde-4ec4-91de-ad1fba9e64a3.png) | ![Screenshot from 2022-11-06 15-21-17](https://user-images.githubusercontent.com/23214006/200180075-05f231d2-977f-4c1c-be4d-ec5efc196b24.png) |

Ray-casting based on the poses from the LLFF dataset (Colmap):

| Example 1 | Example 2 | Example 3 |
| -------------- | -------------- | --------------- |
| ![Screenshot from 2022-11-06 15-25-46](https://user-images.githubusercontent.com/23214006/200180172-d260de6a-0e2f-47c7-8f20-244274a90f4f.png) | ![Screenshot from 2022-11-06 15-24-31](https://user-images.githubusercontent.com/23214006/200180176-d53ec7e0-c073-4e6d-b050-0034e965d4f6.png) | ![Screenshot from 2022-11-06 15-23-47](https://user-images.githubusercontent.com/23214006/200180177-8dc25036-d11d-4a9a-b4a7-30cf0ffb1941.png) |

## What can I do then? (Solution)

It's possible to model only under certain scenarios that may be tedious to accomplish or sometimes may even be impossible. But it can be worth the try.

### Examples of objects that might be feasible to model:

1. Small: Figurine on a table, pendant light, apple hanging on a tree.
2. Large: Building, house, monument, stop sign.

### A few things have to be considered:
- Minimal occlusion (e.g., teacup handle) to none is preferable.
- If something is occluding the object (e.g., a few parts) and you can't do anything. It's worth trying a few things. (e.g., inpainting)
- The generated imagery needs to be rich. (i.e., resolve the whole target object) Use image segmentation (U-Net, Photoshop) and mask everything except the target object - green fill in the referenced image (e.g., air, base). **The objective is to have such data such that the object hangs in a void no matter how you look at it such that the model can parametrize all the space around the object as negative density.**
![Frame 1 (1)](https://user-images.githubusercontent.com/23214006/200187919-be536f19-5fda-43a4-978f-22389a5919e3.png)

- For very large objects, you should have a drone or something similar, which may work only if you're able to achieve accurate pose estimation.
- Objects are not deformable (e.g., humans moving). Although there are a few existing solutions that handle it.
- Have at least 30 shots (from different angles), more is better.
